[DRAFT] Pull proto draft

cprasad1 opened the pull request:
### Description 
_What behavior do you want to change, why, how does your patch achieve the changes?_

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

agavra commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ProtobufQueryStreamResponseWriter.java:
```diff
@@ -0,0 +1,166 @@
+      final byte[] protoMessage = serializer.serialize("", ksqlRecord);
+      final Optional<DataRowProtobuf> protoRow =
+              StreamedRow.pullRowProtobuf(protoMessage).getRowProtobuf();
+      response.write(",\n").write(serializeObject(protoRow));
```
is there any reason we're adding newlines to everything? the other formats seem to just rely on `,` in between rows and then users can pipe it into `jq` or something if they want to

I'm asking because it seems like we may be able to reuse a lot of the `JsonQueryStreamResponseWriter` code if we standardized

cprasad1 replied:
Done. `JsonQueryStreamResponseWriter` now takes care of both json and protobuf format depending on the media type

agavra reviewed (commented):
Did a quick skim to give you some initial feedback, I'll review the tests in more depth tomorrow!

Gerrrr commented:
The new content type behaves differently compared to `application/vnd.ksql.v1+json` for empty responses. We should definitely fix that before merging.

Consider a new table:

```
ksql> CREATE SOURCE TABLE t1 (K INT PRIMARY KEY, VS ARRAY<INT>) WITH (KAFKA_TOPIC='t1', value_format='JSON', partitions=1);
```

Let's query it with `Accept: application/vnd.ksql.v1+protobuf`:

```
curl --header "Accept: application/vnd.ksql.v1+protobuf" \
     -v --request POST \
     -d '{"ksql": "SELECT K, VS FROM t1 WHERE K=2 LIMIT 1;"}' \
      http://localhost:8088/query
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8088 (#0)
> POST /query HTTP/1.1
> Host: localhost:8088
> User-Agent: curl/7.64.1
> Accept: application/vnd.ksql.v1+protobuf
> Content-Length: 51
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 51 out of 51 bytes
< HTTP/1.1 406 Not Acceptable
< content-length: 0
<
* Connection #0 to host localhost left intact
* Closing connection 0
```

We've got `406 Not Acceptable` and an empty response. The same request with `Accept: application/vnd.ksql.v1+json` returns 200 and a header in the body:

```
curl --header "Accept: application/vnd.ksql.v1+json" \
     -v --request POST \
     -d '{"ksql": "SELECT K, VS FROM t1 WHERE K=2 LIMIT 1;"}' \
      http://localhost:8088/query
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8088 (#0)
> POST /query HTTP/1.1
> Host: localhost:8088
> User-Agent: curl/7.64.1
> Accept: application/vnd.ksql.v1+json
> Content-Length: 51
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 51 out of 51 bytes
< HTTP/1.1 200 OK
< Transfer-Encoding: chunked
<
* Connection #0 to host localhost left intact
[{"header":{"queryId":"query_1652258828691","schema":"`K` INTEGER KEY, `VS` ARRAY<INTEGER>"}}]* Closing connection 0
```

Gerrrr commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTarget.java:
```diff
@@ -221,6 +222,21 @@ public RestResponse<List<StreamedRow>> postQueryRequest(
+          final String ksql,
+          final Map<String, ?> requestProperties
+  ) {
+    final Map<String, Object> requestPropertiesObject = requestProperties.entrySet().stream()
```
Can we make `requestProperties` of type `Map<String, Object>`  instead of creating a full copy?

cprasad1 replied:
done

Gerrrr commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTargetUtil.java:
```diff
@@ -69,6 +70,41 @@ public static List<StreamedRow> toRows(final Buffer buff) {
+    boolean headerRead = false;
+    int begin = 0;
+
+    for (int i = 0; i <= buff.length(); i++) {
```
This method deserves a detailed javadoc.

cprasad1 replied:
done

cprasad1 commented:
@Gerrrr the new proto format is only acceptable by the `/query-stream` endpoint and not the `/query` endpoint, so we expect to get a `406` (on all pull queries with the new format and not just on empty tables) while querying the `/query` endpoint.

cprasad1 commented on ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:
```diff
@@ -993,6 +1028,97 @@ public void verify(
+          final String actualSchema = (String) ((Map<?, ?>)actual.get(HEADER_PROTOBUF)).get(SCHEMA);
+          final String expectedSchema = (String) ((Map<?, ?>)expected.get(HEADER_PROTOBUF)).get(SCHEMA);
+
+          assert(actualSchema.equals(expectedSchema));
```
done

cprasad1 commented on ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/StreamedRow.java:
```diff
@@ -197,32 +250,45 @@ public static StreamedRow consistencyToken(final ConsistencyToken consistencyTok
   @JsonCreator
   private StreamedRow(
       @JsonProperty("header") final Optional<Header> header,
+      @JsonProperty("headerProtobuf") final Optional<HeaderProtobuf> headerProtobuf,
```
Done. I have refactored the `Header` to include an `Optional<LogicalSchema> schema` or an `Optional<String> protoSchema` depending on the situation

cprasad1 commented on ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:
```diff
@@ -175,13 +181,27 @@ void buildAndExecuteQuery(final RestTestCase testCase) {
+      List<RqttResponse> protoResponses = ImmutableList.of();
+
+      if (!testCase.getTestLocation().toString().contains("scalable-push-queries")) {
+        protoResponses = statements.queries.stream()
```
done. added a configuration `"testPullWithProtoFormat"` to parameterize the rqtt. When true, it runs and validates the test with the protobuf format

cprasad1 commented on ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:
```diff
@@ -993,6 +1028,97 @@ public void verify(
+      ProtobufSchema schema = null;
+      ProtobufNoSRConverter.Deserializer deserializer = new ProtobufNoSRConverter.Deserializer();
+      final JsonMapper mapper = new JsonMapper();
+      for (int i = 0; i != rows.size(); ++i) {
```
sorry, did not understand this. Can you elaborate?

cprasad1 replied:
If you meant asserting that the header comes before the rows, then I have added an assertion of i==0 down below.

agavra replied:
that's almost what I meant. We should fail if:
- there is no header row
- there is no final row
- the header row doesn't happen at the beginning
- the final row doesn't happen at the end

your new change just addresses the third point

Gerrrr commented:
Thank you for elaborating @cprasad1!

cprasad1 commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlRestClient.java:
```diff
@@ -308,6 +308,14 @@ public RestResponse<List<StreamedRow>> makeQueryRequest(
         ksql, requestPropertiesToSend, Optional.ofNullable(commandSeqNum));
   }
 
+  public RestResponse<List<StreamedRow>> makeQueryStreamRequestProto(
```
All the methods exposed in `KsqlRestClient` and `KsqlTarget` have hard-coded either the endpoint as `/query` or the response format as json or json delimited. We don't have a method that is generic enough to query any combination of endpoints x formats. Ideally, we would want to have an api like that. I tried refactoring all the querying logic into a few generic methods, but it was a bit too much of refactoring given how lots of these parameters are hardcoded and have custom logic wrt sending the query requests (/query and /query-stream have different request frmats), the different content types have different handling and how async handling logic differs from the sync handling logic.

An ideal api would take in (endpoint, content-type, sync/async) as its input, but we don't have that, therefore the need for this

cprasad1 commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTargetUtil.java:
```diff
@@ -69,6 +70,41 @@ public static List<StreamedRow> toRows(final Buffer buff) {
+              rows.add(row);
+              headerRead = true;
+            } else {
+              try {
```
done

cprasad1 commented on ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:
```diff
@@ -993,6 +1028,97 @@ public void verify(
+          try {
+            bytes = mapper.readTree(row.toString()).get("rowProtobuf").get("row").binaryValue();
+          } catch (IOException e) {
+            e.printStackTrace();
```
yep, failing the test upon a json deserialization exception.

cprasad1 commented:
@agavra @Gerrrr Thank you for the helpful comments! :)
I hope that I have addressed all of them. This PR is ready for review again

Gerrrr commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/JsonQueryStreamResponseWriter.java:
```diff
@@ -76,7 +112,23 @@ public QueryStreamResponseWriter writeRow(
       LOG.warn("Dropped tombstone. Not currently supported");
     } else {
-      writeBuffer(ServerUtils.serializeObject(keyValue.value().values()));
+      if (mediaType.equals(KsqlMediaType.KSQL_V1_PROTOBUF.mediaType())) {
```
Do I understand correctly that you merged `JsonQueryStreamResponseWriter` and `ProtobufQueryStreamResponseWriter` since the last review review round and the motivation was to reduce the code duplication as @agavra asked?

There are some differences between `ProtobufQueryStreamResponseWriter` and ` JsonQueryStreamResponseWriter` that were not ported, specifically in methods `writeContinuationToken`, `writeError`, `writeConsistencyToken`, `writeCompletionMessage`, `writeLimitMessage`. Why are they no longer necessary?

Gerrrr replied:
FWIW, I think that it would be better to have format-specific implementations in different classes rather than having conditions in every method. W could reduce code duplication by:
* moving default implementations to the interface or abstract class
* moving format-specfic serialization to a separate class or method.

For completness, we should not necessarily block this PR on it; it is fine with me to just create a follow-up issue.

cprasad1 replied:
addressed

Gerrrr commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTargetUtil.java:
```diff
@@ -69,6 +71,63 @@ public static List<StreamedRow> toRows(final Buffer buff) {
+            .map(row -> toJsonMsg(Buffer.buffer(row), true))
+            .collect(Collectors.toList());
+
+    for (Buffer tidied: buffRows) {
```
Thank you for adding the javadoc and elaborate comments! It helped with understanding this method a lot.

We can further reduce complexity of this method, specifically nesting level and one unnecessary condition, by extracting reading of the header out of the main loop:

```java
    // deserialize the first row (header) into a StreamedRow and add it to
    // the list to return.
    rows.add(deserialize(buffRows.get(0), StreamedRow.class));
    for (int i = 1; i < buffRows.size(); i++) {
      final Buffer tidied = buffRows.get(i);
      try {
        // try to deserialize every subsequent buffer row into a StreamedRow
        // and add it to the list to return. These can be continuation tokens,
        // consistency tokens, error messages, final messages or limit messages.
        // We expect it to fail for the rowProtobuf objects which is handled below.
        rows.add(deserialize(tidied, StreamedRow.class));
      } catch (Exception e) {
        // if it is none of the above, then deserialize it to a
        // StreamedRow.DataRowProtobuf object.
        // We can't directly deserialize into a StreamedRow like we did above
        // because DataRowProtobuf's serialization is {"row":"CAESBlVTRVJfMA=="}
        // whereas the serialization of a StreamedRow containing a DataRowProtobuf
        // looks like {"rowProtobuf":{"row":"CAESBlVTRVJfMA=="}}. The Rest API
        // produces {"row":"CAESBlVTRVJfMA=="}, so we have to do a bit of extra conversion
        // here.
        final DataRowProtobuf protoBytes =
                deserialize(tidied, DataRowProtobuf.class);
        final byte[] bytes = protoBytes.getRow();
        rows.add(StreamedRow.pullRowProtobuf(bytes));
      }
    }
    return rows;
```

WDYT?

cprasad1 replied:
looks great. incorporated the suggestion

Gerrrr replied:
Thank you!

agavra commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTargetUtil.java:
```diff
@@ -69,6 +71,63 @@ public static List<StreamedRow> toRows(final Buffer buff) {
+    final List<Buffer> buffRows = Arrays
+            .stream(buff
+                    .toString()
+                    .replace("},{\"row\":", "},\n\t\n{\"row\":")
+                    .split("\n\t\n"))
```
this seems pretty brittle - I guess we "know" that the contents of `row` are protobuf encoded base64 bytes, right? Is a comma even possible to show up in base64? if it isn't base64, then why can't `\n\t\n` show up as well?

can we use a "real" json parser here to extract each row?

cprasad1 replied:
using a real json parser now

agavra commented on ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/KsqlTargetUtil.java:
```diff
@@ -69,6 +71,63 @@ public static List<StreamedRow> toRows(final Buffer buff) {
+          // looks like {"rowProtobuf":{"row":"CAESBlVTRVJfMA=="}}. The Rest API
+          // produces {"row":"CAESBlVTRVJfMA=="}, so we have to do a bit of extra conversion
+          // here.
+          final DataRowProtobuf protoBytes =
```
You mention that the Rest API produced the StreamdRow directly - what produces `DataRowProtobuf` and can we just standardize on one format? (i.e. why can't they all produce just `{"row":"CAESBlVTRVJfMA=="}`?

cprasad1 replied:
>You mention that the Rest API produced the StreamdRow directly
not 100%. The REST API for proto produces `StreamedRow` for everything except the protobuf data rows. For the protobuf data rows, the `JsonQueryStreamResponseWriter` would produce the `DataRowProtobuf` which is an optional member of `StreamedRow` and gets serialized to `{"row":"CAESBlVTRVJfMA=="}`. 

Why is there this weird conditionality for protobuf data rows? If we just serialize a `StreamedRow` that contains a `DataRowProtobuf`, then it would look something like `{"rowProtobuf":{"row":"CAESBlVTRVJfMA=="}` which is not the public api that we want to abide by (unless we change our minds and loosen the restrictions on the api).

We can achieve the behavior where serializing a `StreamedRow` that contains protobuf row data yields something closer to `{"row":"CAESBlVTRVJfMA=="}`, but we can never get it to be the same because `row` is already taken up by the regular JSON rows. I hope this makes sense

agavra commented on ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/StreamedRow.java:
```diff
@@ -325,25 +377,32 @@ public QueryId getQueryId() {
     private Header(
         final QueryId queryId,
-        final LogicalSchema columnsSchema
+        final Optional<LogicalSchema> columnsSchema,
```
nit: as discussed offline, I think we should just keep this as required

agavra commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/JsonQueryStreamResponseWriter.java:
```diff
@@ -54,16 +72,34 @@ public class JsonQueryStreamResponseWriter implements QueryStreamResponseWriter
+      serializer = new ProtobufNoSRSerdeFactory(ImmutableMap.of())
+              .createSerializer(connectSchema, Struct.class, false);
+
+      final StreamedRow header = StreamedRow.headerProtobuf(
```
just checked here, we have `schema` anyway so it should be pretty cheap to just pipe it in right here

agavra commented on ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/pull-queries-protobuf.json:
```diff
@@ -0,0 +1,288 @@
+        "SELECT * FROM riderLocations LIMIT 5;"
+      ],
+      "topics": [
+        {"name": "test_topic", "partitions": 1} // to get a stable ordering
```
I believe we use a Jackson add on that supports comments

agavra reviewed (commented):
This is getting really close, only two things I consider "blocking" are the questions I asked on `KsqlTargetUtil`

cprasad1 commented on ksqldb-cli/src/main/java/io/confluent/ksql/cli/console/Console.java:
```diff
@@ -406,7 +406,7 @@ private void printRowHeader(final Header header) {
             TabularRow.createHeader(
                 getWidth(),
-                header.getSchema().columns(),
+                header.getSchema().get().columns(),
```
making the schema non-optional, and always including it

cprasad1 commented on ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/RestApiTest.java:
```diff
@@ -916,6 +919,64 @@ public void shouldExecutePullQueryOverHttp2QueryStream() {
+  }
+
+  @Test
+  public void shouldNotExecutePullQueryOverHttp2QueryProto() {
```
This PR would've become harder to manage. I can make a follow-up and enable the same format on `/query` as well. For the purposes of this PR, I would vote for shipping this without touching `/query` to unblock the release.

cprasad1 commented:
addressed blocking comments on `KsqlTargetUtil` and made logical schema non optional in the streamed row header

cprasad1 commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/JsonQueryStreamResponseWriter.java:
```diff
@@ -54,16 +72,34 @@ public class JsonQueryStreamResponseWriter implements QueryStreamResponseWriter
   public QueryStreamResponseWriter writeMetadata(final QueryResponseMetadata metaData) {
     final Buffer buff = Buffer.buffer().appendByte((byte) '[');
-    buff.appendBuffer(ServerUtils.serializeObject(metaData));
+    if (mediaType.equals(KsqlMediaType.KSQL_V1_PROTOBUF.mediaType())) {
```
addressed this. with an enum for the serialization

cprasad1 commented:
I moved most of the serialization logic to `JsonStreamedRowResponseWriter` and refactored it to serialize depending on the response format (proto or json). I also made the proto byte[] and optional field in StreamedRow.DataRow. This also gets rid of the ugly pro parsing logic in `KsqlTargetUtil` since we can now leverage the already existing parser logic there. Ready for review again @agavra @Gerrrr

agavra commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/JsonStreamedRowResponseWriter.java:
```diff
@@ -178,6 +186,72 @@ public void end() {
     response.end();
   }
 
+  public enum RowFormat {
```
nice :)

agavra reviewed (approved):
Thanks @cprasad1! This looks good to me, we can address the minor follow-ups (if any are leftover) in a future PR
